src/main/scala/com/ambiata/poacher/hdfs/HdfsFile.scala

```diff
@@ -0,0 +1,255 @@
+
+  def read: Hdfs[Option[String]] =
+    readWithEncoding(Codec.UTF8)
+
+  def readOrFail: Hdfs[String] =
+    read.flatMap(Hdfs.fromOption(_, s"Failed to read file - HdfsFile($path) does not exist"))
+
+  def readWithEncoding(encoding: Codec): Hdfs[Option[String]] =
+    optionExists(Hdfs.using(toInputStream){ in => Hdfs.fromRIO(Streams.readWithEncoding(in, encoding)) })
+
+  def readLines: Hdfs[Option[List[String]]] =
+    readLinesWithEncoding(Codec.UTF8)
+
+  def readLinesWithEncoding(encoding: Codec): Hdfs[Option[List[String]]] =
+    optionExists(readPerLineWithEncoding(encoding,
+      scala.collection.mutable.ListBuffer[String]())((s, b) => { b += s; b }).map(_.toList))
+
+  def doPerLine[A](f: String => Hdfs[Unit]): Hdfs[Unit] = for {
+    c <- Hdfs.configuration
+    _ <- Hdfs.using(toInputStream)(in =>
+      Hdfs.io {
+        val reader = new java.io.BufferedReader(new java.io.InputStreamReader(in, "UTF-8"))
+        var line: String = null
+        var result: Hdfs[Unit] = null
+        while ({ line = reader.readLine; line != null && result == null })
+          f(line).run(c).unsafePerformIO match {
+            case Ok(_) =>
+              ()
+            case e @ Error(_) =>
+              result = Hdfs.result[Unit](e)
+          }
+        }
+    )
+  } yield ()
+
+  def readPerLine[A](empty: => A)(f: (String, A) => A): Hdfs[A] =
+    readPerLineWithEncoding(Codec.UTF8, empty)(f)
+
+  def readPerLineWithEncoding[A](codec: Codec, empty: => A)(f: (String, A) => A): Hdfs[A] =
+    Hdfs.io(empty).flatMap { s =>
+      var state = s
+      readUnsafe { in => Hdfs.io {
+        val reader = new java.io.BufferedReader(new java.io.InputStreamReader(in, codec.name))
+        var line: String = null
+        while ({ line = reader.readLine; line != null })
+          state = f(line, state)
+      }}.as(state)
+    }
+
+  def readBytes: Hdfs[Option[Array[Byte]]] =
+    optionExists(Hdfs.using(toInputStream)(o => Hdfs.fromRIO(Streams.readBytes(o))))
+
+  def readUnsafe[A](f: InputStream => Hdfs[Unit]): Hdfs[Unit] =
+    Hdfs.using(toInputStream)(f)
+
+  def size: Hdfs[Option[BytesQuantity]] =
+    optionExists(withFileSystem(_.getFileStatus(toHPath).getLen.bytes))
+
+  def sizeOrFail: Hdfs[BytesQuantity] =
+    size.flatMap(Hdfs.fromOption(_, s"Can not calculate the size of a file that does not exist, HdfsFile($path)."))
+
+  def checksum(algorithm: ChecksumAlgorithm): Hdfs[Option[Checksum]] =
+    optionExists(Hdfs.using(toInputStream)(o => Hdfs.fromRIO(Checksum.stream(o, algorithm))))
+
+  def lineCount: Hdfs[Option[Int]] =
+    optionExists(for {
+      i <- toInputStreamReader
+      r <- Hdfs.io({
+        val reader = new LineNumberReader(i)
+        try {
+          while (reader.readLine != null) {}
+            reader.getLineNumber
+        } finally reader.close
+      })
+    } yield r)
+
+  def writeWith[A](f: OutputStream => Hdfs[A]): Hdfs[A] = // not sure about this
```

Comment:
Not sure. This will overwrite the contents of the file, same with `writeStream`. Is that the concern?